sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/state/StateDataSource.scala

```diff
@@ -0,0 +1,216 @@
+      storeName: String,
+      joinSide: JoinSideValues) {
+    def stateCheckpointLocation: Path = new Path(resolvedCpLocation, "state")
+  }
+
+  object StateSourceOptions {
+    def apply(
+        sparkSession: SparkSession,
+        hadoopConf: Configuration,
+        properties: util.Map[String, String]): StateSourceOptions = {
+      apply(sparkSession, hadoopConf, new CaseInsensitiveStringMap(properties))
+    }
+
+    def apply(
+        sparkSession: SparkSession,
+        hadoopConf: Configuration,
+        options: CaseInsensitiveStringMap): StateSourceOptions = {
+      val checkpointLocation = Option(options.get(PARAM_PATH)).orElse {
+        throw new IllegalArgumentException(s"'$PARAM_PATH' must be specified.")
+      }.get
+
+      val resolvedCpLocation = resolvedCheckpointLocation(hadoopConf, checkpointLocation)
+
+      val batchId = Option(options.get(PARAM_BATCH_ID)).map(_.toLong).orElse {
+        Some(getLastCommittedBatch(sparkSession, resolvedCpLocation))
+      }.get
+
+      if (batchId < 0) {
+        throw new IllegalArgumentException(s"'${PARAM_BATCH_ID}' cannot be negative.")
+      }
+
+      val operatorId = Option(options.get(PARAM_OPERATOR_ID)).map(_.toInt)
+        .orElse(Some(0)).get
+
+      if (operatorId < 0) {
+        throw new IllegalArgumentException(s"'${PARAM_OPERATOR_ID}' cannot be negative.")
+      }
+
+      val storeName = Option(options.get(PARAM_STORE_NAME))
+        .getOrElse(StateStoreId.DEFAULT_STORE_NAME)
+
+      if (storeName.isEmpty) {
+        throw new IllegalArgumentException(s"'${PARAM_STORE_NAME}' cannot be an empty string.")
+      }
+
+      val joinSide = try {
+        Option(options.get(PARAM_JOIN_SIDE))
+          .map(JoinSideValues.withName).getOrElse(JoinSideValues.none)
+      } catch {
+        case _: NoSuchElementException =>
+          // convert to IllegalArgumentException
+          throw new IllegalArgumentException(s"Incorrect value of the option " +
+            s"'$PARAM_JOIN_SIDE'. Valid values are ${JoinSideValues.values.mkString(",")}")
+      }
+
+      if (joinSide != JoinSideValues.none && storeName != StateStoreId.DEFAULT_STORE_NAME) {
+        throw new IllegalArgumentException(s"The options '$PARAM_JOIN_SIDE' and " +
+          s"'$PARAM_STORE_NAME' cannot be specified together. Please specify either one.")
+      }
+
+      StateSourceOptions(resolvedCpLocation, batchId, operatorId, storeName, joinSide)
+    }
+
+    private def resolvedCheckpointLocation(
+        hadoopConf: Configuration,
+        checkpointLocation: String): String = {
+      val checkpointPath = new Path(checkpointLocation)
+      val fs = checkpointPath.getFileSystem(hadoopConf)
+      checkpointPath.makeQualified(fs.getUri, fs.getWorkingDirectory).toUri.toString
+    }
+
+    private def getLastCommittedBatch(session: SparkSession, checkpointLocation: String): Long = {
+      val commitLog = new CommitLog(session, new Path(checkpointLocation, "commits").toString)
```

Comment:
can we put `"commits"` and `"state"` as constants into a class like `object StreamingCheckpointConstants` or something?

Reply:
Yeah let's do that. I think this should have been done much earlier, but better late than never.